homeassistant/components/netgear/router.py

```diff
@@ -0,0 +1,287 @@
+        self.firmware_version = self._info["Firmwareversion"]
+
+        if self.model in MODELS_V2:
+            self._method_version = 2
+
+    async def async_setup(self) -> None:
+        """Set up a Netgear router."""
+        await self.hass.async_add_executor_job(self._setup)
+
+        # set already known devices to away instead of unavailable
+        device_registry = dr.async_get(self.hass)
+        devices = dr.async_entries_for_config_entry(device_registry, self.entry_id)
+        for device_entry in devices:
+            if device_entry.via_device_id is None:
+                continue  # do not add the router itself
+
+            device_mac = dict(device_entry.connections).get(dr.CONNECTION_NETWORK_MAC)
+            self.devices[device_mac] = {
+                "mac": device_mac,
+                "name": device_entry.name,
+                "active": False,
+                "last_seen": dt_util.utcnow() - timedelta(days=365),
+                "device_model": None,
+                "device_type": None,
+                "type": None,
+                "link_rate": None,
+                "signal": None,
+                "ip": None,
+            }
+
+        await self.async_update_device_trackers()
+        self._unsub_dispatcher = async_track_time_interval(
+            self.hass, self.async_update_device_trackers, SCAN_INTERVAL
+        )
+
+        async_dispatcher_send(self.hass, self.signal_device_new)
+
+    async def async_unload(self) -> None:
+        """Unload a Netgear router."""
+        self._unsub_dispatcher()
+        self._unsub_dispatcher = None
+
+    async def async_get_attached_devices(self) -> None:
+        """Get the devices connected to the router."""
+        if self._method_version == 1:
+            return await self.hass.async_add_executor_job(
+                self._api.get_attached_devices
+            )
+
+        return await self.hass.async_add_executor_job(self._api.get_attached_devices_2)
+
+    async def async_update_device_trackers(self, now=None) -> None:
+        """Update Netgear devices."""
+        new_device = False
+        ntg_devices = await self.async_get_attached_devices()
+        now = dt_util.utcnow()
+
+        for ntg_device in ntg_devices:
+            device_mac = format_mac(ntg_device.mac)
+
+            if self._method_version == 2 and not ntg_device.link_rate:
+                continue
+
+            if not self.devices.get(device_mac):
+                new_device = True
+
+            # ntg_device is a namedtuple from the collections module that needs conversion to a dict through ._asdict method
+            self.devices[device_mac] = ntg_device._asdict()
+            self.devices[device_mac]["mac"] = device_mac
+            self.devices[device_mac]["last_seen"] = now
+
+        for device in self.devices.values():
+            device["active"] = now - device["last_seen"] <= self._consider_home
+
+        async_dispatcher_send(self.hass, self.signal_device_update)
```

Comment:
yes, because HomeAssistant can be started for a long time and already have setup many devices that are tracked. Then a new device connects to the wifi. That new device will be discovered in the single update call to the router that gives the list of all devices. At that point we want to both update all already existing devices and also signal that a new device is discovered and needs to be setup.